tests/nc-client/main.c

```diff
@@ -0,0 +1,175 @@
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+
+#include "events.h"
+#include "network.h"
+#include "sock.h"
+#include "warnp.h"
+
+struct senddata {
+	char * buffer;
+	ssize_t nchars;
+	int socket;
+	int conndone;
+	void * connect_cookie;
+	void * write_cookie;
+};
+
+/* Forward declarations. */
+static int send_input(void *);
+static int callback_wrote(void * cookie, ssize_t lenwrit);
+static int callback_connected(void * cookie, int socket);
+
+/* Send data from stdin to a socket, or close the connection. */
+static int
+send_input(void * cookie)
+{
+	struct senddata * send = (struct senddata *)cookie;
+	size_t len = 0;
+
+	/* Read data from stdin. */
+	if ((send->nchars = getline(&send->buffer, &len, stdin)) != -1) {
+		/* Send data to server. */
+		if ((send->write_cookie = network_write(send->socket,
+		    (uint8_t *)send->buffer, (size_t)send->nchars,
+		    (size_t)send->nchars, callback_wrote, cookie)) == NULL) {
+			warn0("network_write failure");
+			goto err0;
+		}
+	} else {
+		/* Close connection without sending anything. */
```

Comment:
You might want to check `feof(stdin)` here, to distinguish between EOF and an error?